src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+ *
+ * It plans, manages and executes the cleanup jobs asynchronously in a separate thread so cleaning up does not affect
+ * the performance of the other tasks of the node.
+ */
+public class GarbageCollector {
+    /**
+     * The interval in milliseconds that the garbage collector will check if new cleanup tasks are available.
+     */
+    private static final int GARBAGE_COLLECTOR_RESCAN_INTERVAL = 10000;
+
+    /**
+     * Logger for this class allowing us to dump debug and status messages.
+     */
+    private static final Logger log = LoggerFactory.getLogger(GarbageCollector.class);
+
+    /**
+     * Boolean flag that indicates if the node is being shutdown.
+     */
+    boolean shuttingDown = false;
+
+    /**
+     * Holds a reference to the tangle instance which acts as an interface to the used database.
+     */
+    final Tangle tangle;
+
+    /**
+     * Holds a reference to the {@link SnapshotManager} that this garbage collector belongs to.
+     */
+    final SnapshotManager snapshotManager;
+
+    /**
+     * Holds a reference to the {@link TipsViewModel} which is necessary for removing tips that were pruned.
+     */
+    final TipsViewModel tipsViewModel;
+
+    /**
+     * A map of {@link JobParser}s allowing us to determine how to parse the jobs from the garbage collector state file,
+     * based on their type.
+     */
+    private final HashMap<String, JobParser> jobParsers = new HashMap<>();
+
+    /**
+     * A map of {@link QueueProcessor}s allowing us to process queues based on the type of the job.
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueProcessor> queueProcessors = new HashMap<>();
+
+    /**
+     * A map of {@link QueueConsolidator}s allowing us to consolidate the queues to consume less "space".
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, QueueConsolidator> queueConsolidators = new HashMap<>();
+
+    /**
+     * List of cleanup jobs that shall get processed by the garbage collector (grouped by their class).
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> garbageCollectorJobs = new HashMap<>();
+
+    /**
+     * The constructor of this class stores the passed in parameters for future use and restores the previous state of
+     * the garbage collector if there is a valid one (to continue with cleaning up after IRI restarts).
+     *
+     * Before restoring the garbage collector state it registers the available job types, so we know how to parse and
+     * process the found jobs.
+     */
+    public GarbageCollector(Tangle tangle, SnapshotManager snapshotManager, TipsViewModel tipsViewModel) {
+        this.tangle = tangle;
+        this.snapshotManager = snapshotManager;
+        this.tipsViewModel = tipsViewModel;
+
+        MilestonePrunerJob.registerInGarbageCollector(this);
+        UnconfirmedSubtanglePrunerJob.registerInGarbageCollector(this);
+
+        restoreCleanupJobs();
```

Comment:
I kinda not like this being in the consturctor.
2 main reasons:
1. I consider it unclean to have all this logic in the constructor
2. More practical reason: when you instancize this class in a unit test, you will have to read a file? Maybe you should start taking time writing unit tests.
3. If an exception will be thrown during construction it will violate section 7-3 of [Java Secure Coding Guideline](http://www.oracle.com/technetwork/java/seccodeguide-139067.html#7)